config/file_config.go

```diff
@@ -92,336 +111,137 @@ type PrometheusMetricsConfig struct {
-	case "resize", "impact":
-		break
-	default:
-		return fmt.Errorf("invalid CacheOverrunStrategy: '%s'", st)
-	}
-	return nil
-}
-
-func (f *fileConfig) validateSamplerConfigs() error {
-	logrus.Debugf("Sampler rules config: %+v", f.rules)
-
-	keys := f.rules.AllKeys()
-	for _, key := range keys {
-		parts := strings.Split(key, ".")
-
-		// verify default sampler config
-		if parts[0] == "sampler" {
-			t := f.rules.GetString(key)
-			var i interface{}
-			switch t {
-			case "DeterministicSampler":
-				i = &DeterministicSamplerConfig{}
-			case "DynamicSampler":
-				i = &DynamicSamplerConfig{}
-			case "EMADynamicSampler":
-				i = &EMADynamicSamplerConfig{}
-			case "RulesBasedSampler":
-				i = &RulesBasedSamplerConfig{}
-			case "TotalThroughputSampler":
-				i = &TotalThroughputSamplerConfig{}
-			default:
-				return fmt.Errorf("Invalid or missing default sampler type: %s", t)
-			}
-			err := f.rules.Unmarshal(i)
+func (f *fileConfig) monitor() {
+	f.done = make(chan struct{})
+	f.ticker = time.NewTicker(time.Duration(f.mainConfig.ConfigReloadInterval))
+	for {
+		select {
+		case <-f.done:
+			return
+		case <-f.ticker.C:
+			// reread the configs
+			cfg, err := newConfig(f.opts)
 			if err != nil {
-				return fmt.Errorf("Failed to unmarshal sampler rule: %w", err)
+				f.errorCallback(err)
+				continue
 			}
-			v := validator.New()
-			err = v.Struct(i)
-			if err != nil {
-				return fmt.Errorf("Failed to validate sampler rule: %w", err)
-			}
-		}
 
-		// verify dataset sampler configs
-		if len(parts) > 1 && parts[1] == "sampler" {
-			t := f.rules.GetString(key)
-			var i interface{}
-			switch t {
-			case "DeterministicSampler":
-				i = &DeterministicSamplerConfig{}
-			case "DynamicSampler":
-				i = &DynamicSamplerConfig{}
-			case "EMADynamicSampler":
-				i = &EMADynamicSamplerConfig{}
-			case "RulesBasedSampler":
-				i = &RulesBasedSamplerConfig{}
-			case "TotalThroughputSampler":
-				i = &TotalThroughputSamplerConfig{}
-			default:
-				return fmt.Errorf("Invalid or missing dataset sampler type: %s", t)
+			// if nothing's changed, we're fine
+			if f.mainHash == cfg.mainHash && f.rulesHash == cfg.rulesHash {
+				continue
 			}
-			datasetName := parts[0]
-			if sub := f.rules.Sub(datasetName); sub != nil {
-				err := sub.Unmarshal(i)
-				if err != nil {
-					return fmt.Errorf("Failed to unmarshal dataset sampler rule: %w", err)
-				}
-				v := validator.New()
-				err = v.Struct(i)
-				if err != nil {
-					return fmt.Errorf("Failed to validate dataset sampler rule: %w", err)
-				}
+
+			// otherwise, update our state and call the callbacks
+			f.mux.Lock()
+			f.mainConfig = cfg.mainConfig
+			f.mainHash = cfg.mainHash
+			f.rulesConfig = cfg.rulesConfig
+			f.rulesHash = cfg.rulesHash
+			for _, cb := range f.callbacks {
+				cb()
 			}
+			f.mux.Unlock() // can't defer since the goroutine never ends
 		}
 	}
-	return nil
+}
+
+// Stop halts the monitor goroutine
+func (f *fileConfig) Stop() {
```

Comment:
Is this used anywhere?

Reply:
Not yet, but I think it will be as I further refactor things to allow a cleaner shutdown. Whenever I add a channel I want a Stop capability to be available.